crates/mfem/src/lib.rs

```diff
@@ -1,252 +1,1251 @@
+            c_int(nelem as i32),
+            c_int(nbdr_elem as i32),
+            c_int(space_dim as i32),
+        ));
+        MeshBuilder {
+            inner,
+            nvert,
+            nvert_provided: 0,
+        }
     }
 
+    /// Return a mesh created by reading a file in MFEM, Netgen, or
+    /// VTK format.
+    #[must_use]
     pub fn from_file(path: &str) -> Result<Self, MfemError> {
-        let generate_edges = 1;
-        let refine = 1;
+        let generate_edges = c_int(1);
+        let refine = c_int(1);
         let fix_orientation = true;
         let_cxx_string!(mesh_path = path);
-        let inner =
-            mfem_sys::ffi::Mesh_ctor_file(&mesh_path, generate_edges, refine, fix_orientation);
-        Ok(Self { inner })
+        Ok(Self::emplace(mfem_sys::Mesh::new6(
+            &mesh_path,
+            generate_edges,
+            refine,
+            fix_orientation,
+        )))
     }
 
+    #[must_use]
     pub fn dimension(&self) -> i32 {
-        self.inner.Dimension()
+        self.as_mfem().Dimension().into()
     }
 
+    #[must_use]
     pub fn get_num_elems(&self) -> i32 {
-        self.inner.GetNE()
+        self.as_mfem().GetNE().into()
     }
 
-    pub fn get_nodes<'fes, 'a: 'fes>(&'a self) -> Option<GridFunctionRef<'fes, 'a>> {
-        mfem_sys::ffi::Mesh_GetNodes(&self.inner)
-            .ok()
-            .map(|grid_func| GridFunctionRef { inner: grid_func })
+    pub fn get_nodes<'a>(&'a self) -> Option<Ref<'a, GridFunction<'a>>> {
+        let nodes = self.as_mfem().GetNodes2();
+        if nodes.is_null() {
+            None
+        } else {
+            Some(Ref::from_ptr(nodes))
+        }
     }
 
-    pub fn get_bdr_attributes<'a>(&'a self) -> ArrayIntRef<'a> {
-        let inner = mfem_sys::ffi::Mesh_bdr_attributes(&self.inner);
-        ArrayIntRef { inner }
+    pub fn bdr_attributes(&self) -> Ref<'_, ArrayInt> {
+        let attr = mfem_sys::Mesh_bdr_attributes(self.as_mfem());
+        Ref::from_ref(attr)
     }
 
     pub fn uniform_refinement(&mut self, ref_algo: RefAlgo) {
-        self.inner.pin_mut().UniformRefinement(ref_algo as i32);
+        self.as_mfem_mut()
+            .UniformRefinement1(c_int(ref_algo as i32))
     }
 
-    pub fn save_to_file(&self, path: &str, precision: i32) {
+    pub fn save(&self) -> MeshSave<'_> {
+        MeshSave {
+            mesh: self,
+            precision: 8,
+        }
+    }
+
+    /// Add a [`FiniteElementCollection`] to the mesh `self`.
+    ///
+    /// The argument of the function `fec` is the `FiniteElementCollection`
+    /// associated to the nodes of the mesh, if any.  This collection
+    /// is used to represent the curvature of the elements (that is,
+    /// how the geometry of the problem is discretized).  The
+    /// `FiniteElementCollection` returned by `fec` is the one that
+    /// will be used to discretize the PDE.
+    #[must_use]
+    pub fn with_fec<'a, FEC>(&'a mut self, fec: FEC) -> MeshWithFEC<'a>
```

Comment:
```suggestion
    pub fn map_fec<'a, FecFunc>(&'a mut self, f: FecFunc) -> MeshWithFEC<'a>
```
The name `with_fec` feels odd, because all other `with_$X` functions are part of a "builder pattern". This one is the odd one out, for two main reasons.

First, all other "builder pattern" functions return an entirely independent new thing. This one returns something that holds a reference to the original.

Second, all other "builder pattern" functions take a plain-old-data value that is meant to replace the original value, without regard to what it was. This one processes the original value to produce the new value.

Mostly for the second reason, a name of `map_$X` form seems easier to understand.